src/blas/impl/KokkosBlas2_gemv_impl.hpp

```diff
@@ -480,6 +481,132 @@ singleLevelGemv (const char trans[],
+
+  TwoLevelGEMV (const AlphaCoeffType& alpha,
+                         const AViewType& A,
+                         const XViewType& x,
+                         const BetaCoeffType& beta,
+                         const YViewType& y) :
+    alpha_ (alpha), A_ (A), x_ (x), beta_ (beta), y_ (y)
+  {
+    static_assert (Kokkos::Impl::is_view<AViewType>::value,
+                   "AViewType must be a Kokkos::View.");
+    static_assert (Kokkos::Impl::is_view<XViewType>::value,
+                   "XViewType must be a Kokkos::View.");
+    static_assert (Kokkos::Impl::is_view<YViewType>::value,
+                   "YViewType must be a Kokkos::View.");
+    static_assert (static_cast<int> (AViewType::rank) == 2,
+                   "AViewType must have rank 2.");
+    static_assert (static_cast<int> (XViewType::rank) == 1,
+                   "XViewType must have rank 1.");
+    static_assert (static_cast<int> (YViewType::rank) == 1,
+                   "YViewType must have rank 1.");
+    static_assert (std::is_integral<IndexType>::value,
+                   "IndexType must be an integer.");
+  }
+
+public:
+  //LayoutLeft version: 32xK blocks.
+  //  -Each team handles block rows. 
+  //  -Groups of 32 threads handle N/teamsize columns sequentially, placing results into shared.
+  //  -Then individual thread results are combined with parallel_reduce.
+  KOKKOS_INLINE_FUNCTION void
+  operator () (TwoLevelGEMV_LayoutLeftTag, const member_type& team) const
+  {
+    using Kokkos::Details::ArithTraits;
+    using Scalar = typename YViewType::non_const_value_type;
+    using KAT = ArithTraits<Scalar>;
+    //Allocate a Scalar in shared for each thread
+    Scalar* blockResult = (Scalar*) team.team_shmem().get_shmem(32 * sizeof(Scalar));
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(team, 32),
+    [&](int i)
+    {
+      blockResult[i] = KAT::zero();
+    });
+    team.team_barrier();
+    //Which block this thread will work on
+    int block = team.team_rank() / 32;
+    //Which row in the block this thread will work on
+    IndexType row = team.league_rank() * 32 + team.team_rank() % 32;
+    IndexType blockColStart = columnsPerThread * block;
+    Scalar localSum = KAT::zero();
+    //compute local sum
+    for(IndexType col = blockColStart; col < blockColStart + columnsPerThread; col++)
+    {
+      if(col == (IndexType) A_.extent(1))
+        break;
+      if(row < (IndexType) A_.extent(0))
+      {
+        //A access is coalesced, x access is a broadcast
+        localSum += A_(row, col) * x_(col);
+      }
+    }
+    //atomically combine local result into shared
+    Kokkos::atomic_add(&blockResult[team.team_rank() % 32], localSum);
+    team.team_barrier();
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(team, 32),
+    [&](int i)
+    {
+      IndexType yrow = team.league_rank() * 32 + i;
+      if(yrow < (IndexType) A_.extent(0))
+      {
+        y_[yrow] = beta_ * y_[yrow] + alpha_ * blockResult[i];
```

Comment:
Any reason that you use a ``[]`` operator instead of ``()`` ?

Reply:
Nope, good catch.

Reply:
Oh, I think I copied that from the existing TwoLevelTransposeGEMV. I think I'll go through the file and change them all, since right now it's mixing ``()`` and ``[]`` even for vectors.